File: matlab/SPmerge01linear.m
function [sMerge] = SPmerge01linear(scanAngles,varargin)

% Colin Ophus, National Center for Electron Microscopy, Molecular Foundry,
% Lawrence Berkeley National Laboratory, Berkeley, CA, USA. (Feb 2016).

% New version of SPmerge01.m - This script now searches over linear drift
% vectors, aligned to first two images.  This search is performed twice.

% Merge multiple scanning probe images.  Assume scan origin is upper left
% corner of the image, and that the scan direction for zero degrees is
% horizontal (along MATLAB columns).  All input images must have fast scan
% direction along the array rows (horizontal direction).  Original data is 
% stored in 3D arrray sMerge.scanLines

% Inputs:
% scanAngles            - Vector containg scan angles in degrees.
% images                - 2D image arrays, order in scanAngles, all same size.
% sigmaLP = 32; %       - sigma value in pixels for initial alignment LP filter. 
flagReportProgress = 1;  % Set to true to see updates on console.v =
paddingScale = (1+1/4);%1.5;    % - padding amount for scaling of the output.
sMerge.KDEsigma = 1/2; % - Smoothing between pixels for KDE.
sMerge.edgeWidth = 1/128; % - size of edge blending relative to input images.
sMerge.linearSearch = linspace(-0.02,0.02,1+2*2);  % Initial linear search vector, relative to image size.
% sMerge.linearSearch = linspace(-0.04,0.04,1+2*4);  % Initial linear search vector, relative to image size.

% flagSkipInitialAlignment = 0;  % Set to true to skip initial phase correlation.
% flagCrossCorrelation = 0+1;  % Set to true to use cross correlation.
% sigmaAlignLP = 64;  % low pass filtering for initial alignment (pixels).

% Initialize struct containing all data
sMerge.imageSize = round([size(varargin{1},1) size(varargin{1},2)]*paddingScale/4)*4;
sMerge.scanAngles = scanAngles;
sMerge.numImages = length(scanAngles);
sMerge.scanLines = zeros(size(varargin{1},1),size(varargin{1},2),sMerge.numImages);
sMerge.scanOr = zeros(size(varargin{1},1),2,sMerge.numImages);
sMerge.scanDir = zeros(sMerge.numImages,2);
sMerge.imageTransform = zeros(sMerge.imageSize(1),sMerge.imageSize(2),sMerge.numImages);
sMerge.imageDensity = zeros(sMerge.imageSize(1),sMerge.imageSize(2),sMerge.numImages);

% Generate all initial variables 
for a0 = 1:sMerge.numImages
    % Raw images
    if nargin == (sMerge.numImages+1)
        % Sequential images
        sMerge.scanLines(:,:,a0) = varargin{a0};
    elseif nargin == 2
        % 3D image stack
         sMerge.scanLines(:,:,a0) = varargin{1}(:,:,a0);
    else
        
    end
    
    % Generate pixel origins
    if nargin == (sMerge.numImages+1)
        xy = [(1:size(varargin{a0},1))' ones(size(varargin{a0},1),1)];
        xy(:,1) = xy(:,1) - size(varargin{a0},1)/2;
        xy(:,2) = xy(:,2) - size(varargin{a0},2)/2;
    elseif nargin == 2
        xy = [(1:size(varargin{1}(:,:,a0),1))' ones(size(varargin{1}(:,:,a0),1),1)];
        xy(:,1) = xy(:,1) - size(varargin{1}(:,:,a0),1)/2;
        xy(:,2) = xy(:,2) - size(varargin{1}(:,:,a0),2)/2;
    end
    xy = [xy(:,1)*cos(scanAngles(a0)*pi/180) ...
        - xy(:,2)*sin(scanAngles(a0)*pi/180) ...
        xy(:,2)*cos(scanAngles(a0)*pi/180) ...
        + xy(:,1)*sin(scanAngles(a0)*pi/180)];
    xy(:,1) = xy(:,1) + sMerge.imageSize(1)/2;
    xy(:,2) = xy(:,2) + sMerge.imageSize(2)/2;
    xy(:,1) = xy(:,1) - mod(xy(1,1),1);
    xy(:,2) = xy(:,2) - mod(xy(1,2),1);
    sMerge.scanOr(:,:,a0) = xy;

    % Scan direction
    sMerge.scanDir(a0,:) = [cos(scanAngles(a0)*pi/180 + pi/2) ...
        sin(scanAngles(a0)*pi/180 + pi/2)];
end

if flagReportProgress == true
    reverseStr = ''; % initialize console message piece
end


% First linear alignment, search over possible linear drift vectors.
sMerge.linearSearch = sMerge.linearSearch * size(sMerge.scanLines,1);
[yDrift,xDrift] = meshgrid(sMerge.linearSearch);
sMerge.linearSearchScore1 = zeros(length(sMerge.linearSearch));
inds = linspace(-0.5,0.5,size(sMerge.scanLines,1))';
N = size(sMerge.scanLines);
w2 = circshift(padarray(hanningLocal(N(1))*hanningLocal(N(2))',...
    sMerge.imageSize-N(1:2),0,'post'),round((sMerge.imageSize-N(1:2))/2));
% IcorrNorm = ifft2(abs(fft2(w2)).^2,'symmetric');
for a0 = 1:length(sMerge.linearSearch)
    for a1 = 1:length(sMerge.linearSearch)
        % Calculate time dependent lienar drift
        xyShift = [inds*xDrift(a0,a1) inds*yDrift(a0,a1)];
        
        % Apply linear drift to first two images
        sMerge.scanOr(:,:,1:2) = sMerge.scanOr(:,:,1:2) ...
            + repmat(xyShift,[1 1 2]);
        % Generate trial images
        sMerge = SPmakeImage(sMerge,1);
        sMerge = SPmakeImage(sMerge,2);
        % Measure alignment score with hybrid correlation
        m = fft2(w2.*sMerge.imageTransform(:,:,1)) ...
            .* conj(fft2(w2.*sMerge.imageTransform(:,:,2)));
        Icorr = ifft2(sqrt(abs(m)).*exp(1i*angle(m)),'symmetric');
        sMerge.linearSearchScore1(a0,a1) = max(Icorr(:));
        % Remove linear drift from first two images
        sMerge.scanOr(:,:,1:2) = sMerge.scanOr(:,:,1:2) ...
            - repmat(xyShift,[1 1 2]);
        if flagReportProgress == true
            comp = (a1 / length(sMerge.linearSearch) ...
                + a0 - 1) / length(sMerge.linearSearch);
            msg = sprintf(['Initial Linear Drift Search = ' ...
                sprintf('%.02f',100*comp) ' percent complete']);
            fprintf([reverseStr, msg]);
            reverseStr = repmat(sprintf('\b'),1,length(msg));
        end
    end
end


% Second linear alignment, refine possible linear drift vectors.
[~,ind] = max(sMerge.linearSearchScore1(:));
[xInd,yInd] = ind2sub(size(sMerge.linearSearchScore1),ind);
% if xInd == 1
%     xInds = 1:2;
% elseif xInd == size(sMerge.linearSearchScore1,1)
%     xInds = (-1:0) + size(sMerge.linearSearchScore1,1);
% else
%     if sMerge.linearSearchScore1(xInd-1,yInd) ...
%             > sMerge.linearSearchScore1(xInd+1,yInd)
%         xInds = xInd + (-1:0);
%     else
%         xInds = xInd + (0:1);
%     end
% end
% if yInd == 1
%     yInds = 1:2;
% elseif yInd == size(sMerge.linearSearchScore1,2)
%     yInds = (-1:0) + size(sMerge.linearSearchScore1,2);
% else
%     if sMerge.linearSearchScore1(xInd,yInd-1) ...
%             > sMerge.linearSearchScore1(xInd,yInd+1)
%         yInds = yInd + (-1:0);
%     else
%         yInds = yInd + (0:1);
%     end
% end
step = sMerge.linearSearch(2) - sMerge.linearSearch(1);
xRefine = sMerge.linearSearch(xInd) ...
    + linspace(-0.5,0.5,length(sMerge.linearSearch))*step;
yRefine = sMerge.linearSearch(yInd) ...
    + linspace(-0.5,0.5,length(sMerge.linearSearch))*step;
[yDrift,xDrift] = meshgrid(yRefine,xRefine);
sMerge.linearSearchScore2 = zeros(length(sMerge.linearSearch));
for a0 = 1:length(sMerge.linearSearch)
    for a1 = 1:length(sMerge.linearSearch)
        % Calculate time dependent lienar drift
        xyShift = [inds*xDrift(a0,a1) inds*yDrift(a0,a1)];
        
        % Apply linear drift to first two images
        sMerge.scanOr(:,:,1:2) = sMerge.scanOr(:,:,1:2) ...
            + repmat(xyShift,[1 1 2]);
        % Generate trial images
        sMerge = SPmakeImage(sMerge,1);
        sMerge = SPmakeImage(sMerge,2);
        % Measure alignment score with hybrid correlation
        m = fft2(w2.*sMerge.imageTransform(:,:,1)) ...
            .* conj(fft2(w2.*sMerge.imageTransform(:,:,2)));
        Icorr = ifft2(sqrt(abs(m)).*exp(1i*angle(m)),'symmetric');
        sMerge.linearSearchScore2(a0,a1) = max(Icorr(:));
        % Remove linear drift from first two images
        sMerge.scanOr(:,:,1:2) = sMerge.scanOr(:,:,1:2) ...
            - repmat(xyShift,[1 1 2]);
        if flagReportProgress == true
            comp = (a1 / length(sMerge.linearSearch) ...
                + a0 - 1) / length(sMerge.linearSearch);
            msg = sprintf(['Linear Drift Refinement = ' ...
                sprintf('%.02f',100*comp) ' percent complete']);
            fprintf([reverseStr, msg]);
            reverseStr = repmat(sprintf('\b'),1,length(msg));
        end
    end
end
[~,ind] = max(sMerge.linearSearchScore2(:));
[xInd,yInd] = ind2sub(size(sMerge.linearSearchScore2),ind);
sMerge.xyLinearDrift = [xDrift(xInd) yDrift(yInd)];


% Apply linear drift to all images
xyShift = [inds*sMerge.xyLinearDrift(1) inds*sMerge.xyLinearDrift(2)];
for a0 = 1:sMerge.numImages
    sMerge.scanOr(:,:,a0) = sMerge.scanOr(:,:,a0) + xyShift;
    sMerge = SPmakeImage(sMerge,a0);
end

% Estimate initial alignment
dxy = zeros(sMerge.numImages,2);
G1 = fft2(w2.*sMerge.imageTransform(:,:,1));
for a0 = 2:sMerge.numImages
    G2 = fft2(w2.*sMerge.imageTransform(:,:,a0));
    m = G1.*conj(G2);
    Icorr = ifft2(sqrt(abs(m)).*exp(1i*angle(m)),'symmetric');
    [~,ind] = max(Icorr(:));
    [dx,dy] = ind2sub(size(Icorr),ind);
    dx = mod(dx - 1 + size(Icorr,1)/2,size(Icorr,1)) - size(Icorr,1)/2;
    dy = mod(dy - 1 + size(Icorr,2)/2,size(Icorr,2)) - size(Icorr,2)/2;
    dxy(a0,:) = dxy(a0-1,:) + [dx dy];
    G1 = G2;
end
dxy(:,1) = dxy(:,1) - mean(dxy(:,1));
dxy(:,2) = dxy(:,2) - mean(dxy(:,2));
% Apply alignments and regenerate images
for a0 = 1:sMerge.numImages
    sMerge.scanOr(:,1,a0) = sMerge.scanOr(:,1,a0) + dxy(a0,1);
    sMerge.scanOr(:,2,a0) = sMerge.scanOr(:,2,a0) + dxy(a0,2);
    sMerge = SPmakeImage(sMerge,a0);
end
% Set reference point
sMerge.ref = round(sMerge.imageSize/2);


% Plot results, image with scanline origins overlaid
imagePlot = mean(sMerge.imageTransform,3);
dens = prod(sMerge.imageDensity,3);
% Scale intensity of image
mask = dens>0.5;
imagePlot = imagePlot - mean(imagePlot(mask));
imagePlot = imagePlot / sqrt(mean(imagePlot(mask).^2));

figure(1)
clf
imagesc(imagePlot)
hold on
cvals = [1 0 0;
    0 .7 0;
    0 .6 1;
    1 .7 0;
    1 0 1;
    0 0 1];
for a0 = 1:sMerge.numImages
    scatter(sMerge.scanOr(:,2,a0),sMerge.scanOr(:,1,a0),'marker','.',...
        'sizedata',25,'markeredgecolor',cvals(mod(a0-1,size(cvals,1))+1,:))
end
scatter(sMerge.ref(2),sMerge.ref(1),...
    'marker','+','sizedata',500,...
    'markeredgecolor',[1 1 0],'linewidth',6)
scatter(sMerge.ref(2),sMerge.ref(1),...
    'marker','+','sizedata',500,...
    'markeredgecolor','k','linewidth',2)
hold off
axis equal off
colormap(gray(256))
set(gca,'position',[0 0 1 1])
caxis([-3 3])   % units of image RMS 

if flagReportProgress == true
    fprintf([reverseStr ' ']);
end
end

% function [Iscale] = scaleImage(I,intRange,sigmaLP)
% Iscale = I - mean(I(:));
% Iscale = Iscale / sqrt(mean(I(:).^2));
% Iscale = (Iscale - intRange(1))/(intRange(2)-intRange(1));
% if sigmaLP > 0
%     sm = fspecial('gaussian',2*ceil(3*sigmaLP)+1,sigmaLP);
%     Iscale = conv2(Iscale,sm,'same') ...
%         ./ conv2(ones(size(Iscale)),sm,'same');
% end
% Iscale(Iscale<0) = 0;
% Iscale(Iscale>1) = 1;
% end

function [w] = hanningLocal(N)
% Replacement for simplest 1D hanning function to remove dependency.
w = sin(pi*((1:N)'/(N+1))).^2;
end

File: python/scanning_drift_corr/SPmerge01linear.py
"""The file contains the SPmerge01linear function
"""

import warnings
from multiprocessing import Pool, cpu_count

import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm

from scanning_drift_corr.sMerge import sMerge
from scanning_drift_corr.SPmakeImage import SPmakeImage, makeImage
from scanning_drift_corr.tools import hybrid_correlation

def SPmerge01linear(scanAngles, *images, **kwargs):
    """
    New version of SPmerge01.m - This script now searches over linear drift
    vectors, aligned to first two images.  This search is performed twice.

    Merge multiple scanning probe images.  Assume scan origin is upper left
    corner of the image, and that the scan direction for zero degrees is
    horizontal (along MATLAB columns).  All input images must have fast scan
    direction along the array rows (horizontal direction).  Original data is
    stored in 3D arrray sMerge.scanLines

    Parameters
    ----------
    scanAngles : array-like
        the scan angles in degrees, the same order as provided images.
    images : array-like
        provided images, can be a sequence of images (e.g. img1, img2, img3,
        im4) or as a stack (three-dimensional structure with navigation index
        first).
    linearSearch : array-like
        a 1 dimensional array contains the initial guess for the alignment.
        The default is [-0.02, -0.01,  0.  ,  0.01,  0.02].
    paddingScale : float, optional
        the scale to pad around the images, this cannot be too small
        (i.e. too close to 1). The default is 1.125.
    niter : int, optional
        the number of linear drift search to be performed. The default is 2.
    parallel : bool, optional
        whether to parallelise the search of linear drifts.
        The default is True.
    flagReportProgress : bool, optional
        whether to show progress bars or not. Default to True.
    flagPlot : bool, optional
        whether to show plot after linear drift correction. The default is True.

    Returns
    -------
    sm : sMerge object
        the sMerge object contains all the data.
    """

    # do nothing if no image is provided
    if len(images) == 0:
        return

    # ignore unknown input arguments
    _args_list = ['linearSearch', 'paddingScale', 'flagReportProgress',
                  'parallel', 'flagPlot', 'niter']
    for key in kwargs.keys():
        if key not in _args_list:
            msg = "The argument '{}' is not recognised, and it is ignored."
            warnings.warn(msg.format(key), RuntimeWarning)

    # set default values or from input arguments
    linearSearch = np.asarray(kwargs.get('linearSearch',
                                         np.linspace(-0.02, 0.02, num=2*2+1)))
    paddingScale = kwargs.get('paddingScale', 1.125)
    flagReportProgress = kwargs.get('flagReportProgress', True)
    flagPlot = kwargs.get('flagPlot', True)
    parallel = kwargs.get('parallel', True)
    niter = kwargs.get('niter', 2)

    # initialise the sMerge object
    scanAngles = np.asarray(scanAngles)
    sm = sMerge(scanAngles, images,  paddingScale=paddingScale)

    # get testing linear drifts
    linearSearch *= sm.nr
    inds = np.linspace(-0.5, 0.5, num=sm.nr)[:, None]

    # get linear drift, using the first two images
    xdrift, ydrift = _get_linear_drift(sm, linearSearch, flagReportProgress,
                                       parallel, inds, niter)
    sm.xyLinearDrift = np.array([xdrift, ydrift])

    # apply linear drift to all images
    xyShift = np.hstack([inds*xdrift, inds*ydrift]).T
    _shift_origins(sm, xyShift[None, ...])

    # estimate initial alignment
    dxy = _get_initial_alignment(sm)

    # apply alignments and regenerate images
    _shift_origins(sm, dxy[..., None])

    if flagPlot:
        _plot(sm)

    return sm

def _get_linear_drift(sm, linearSearch, flagReportProgress, parallel, inds,
                      niter=2):

    # matrix to store all correlation scores
    scores = np.empty((niter, linearSearch.size, linearSearch.size))

    # drift the image with testing drifts, refine along the way
    xRefine, yRefine = linearSearch, linearSearch
    for k in range(niter - 1):
        # get alignment score for specific drifts, first the linearSearch
        # then the refined value
        score = _get_linear_alignment_score(sm, linearSearch, inds,
                                            flagReportProgress, parallel,
                                            xRefine, yRefine)

        # record the score for this set of drift
        scores[k, ...] = score

        # refine the size of drift
        xInd, yInd = np.unravel_index(score.argmax(), score.shape)
        xstep = np.diff(xRefine)[0]
        ystep = np.diff(yRefine)[0]
        xRefine = xRefine[xInd] + np.linspace(-0.5, 0.5, xRefine.size)*xstep
        yRefine = yRefine[yInd] + np.linspace(-0.5, 0.5, yRefine.size)*ystep

    # get final score
    score = _get_linear_alignment_score(sm, linearSearch, inds,
                                        flagReportProgress, parallel,
                                        xRefine, yRefine)
    scores[niter-1, ...] = score
    sm.linearSearchScores = scores

    # get the drifts, xInd and yInd are the indices of drfits resulting
    # in the highest linearSearchScore
    xInd, yInd = np.unravel_index(score.argmax(), score.shape)
    xDrift, yDrift = np.meshgrid(xRefine, yRefine)

    r1, c1 = np.unravel_index(xInd, xDrift.shape)
    r2, c2 = np.unravel_index(yInd, yDrift.shape)
    xdrift = xDrift[r1, c1]
    ydrift = yDrift[r2, c2]

    return xdrift, ydrift

def _get_linear_alignment_score(sm, linearSearch, inds, flagReportProgress,
                                parallel, xcoord=None, ycoord=None):
    """Perform linear alignment

    Parameters
    ----------
    xcoord : array-like, optional
        the candidates for x drifts. The default is None, set to the
        default linearSearch.
    ycoord : array-like, optional
        the candidates for y drifts. The default is None, set to the
        default linearSearch.

    Returns
    -------
    linearSearchScore : ndarray
        the array storing the score for correlation for different drifts.
    """

    # xcoord, ycoord set the value/size of meshgrid
    if xcoord is None:
        xcoord = linearSearch
    else:
        xcoord = np.asarray(xcoord)

    if ycoord is None:
        ycoord = linearSearch
    else:
        ycoord = np.asarray(ycoord)

    # construct x and y drifts and initialise score array
    yDrift, xDrift = np.meshgrid(ycoord, xcoord)
    linearSearchScore = np.zeros((linearSearch.size, linearSearch.size))

    if parallel:
        # the tasks include all shifted scanline origins by different shifts
        tasks = []
        for a0 in range(linearSearch.size):
            for a1 in range(linearSearch.size):
                xyShift = np.hstack([inds*xDrift[a0,a1], inds*yDrift[a0,a1]])
                shiftedOr = sm.scanOr[:2, ...] + xyShift.T
                tasks.append([shiftedOr, a0, a1])

        # set global variables to avoid duplication in every worker
        _set_global_sMerge_obj(sm)

        linearSearchScore = _parallel_search(linearSearch,
                                             flagReportProgress, tasks)
    else:
        # perform serial search
        linearSearchScore = _serial_search(sm, linearSearch,
                                           flagReportProgress, inds,
                                           xDrift, yDrift)

    return linearSearchScore

def _set_global_sMerge_obj(sm):
    """Do not want to copy duplicate data in each worker
    Make global variables for the workers to use
    """
    global Gscanline01, GscanDir01, GimageSize, GKDEsigma, Gimg_shape

    Gscanline01 = sm.scanLines[:2, ...]
    GscanDir01 = sm.scanDir[:2, :]
    GimageSize = sm.imageSize
    GKDEsigma = sm.KDEsigma
    Gimg_shape = sm.img_shape

    return

def _makeimage(task):
    """determine the correlation score for this particular shifted origins
    a0, a1 records the index of the drift
    """

    shiftedScanOr, a0, a1 = task

    # generate trial images after applying specific drifts
    img0 = makeImage(Gscanline01[0,...], shiftedScanOr[0,...],
                     GscanDir01[0,:], GimageSize, GKDEsigma)
    img1 = makeImage(Gscanline01[1,...], shiftedScanOr[1,...],
                     GscanDir01[1,:], GimageSize, GKDEsigma)

    # measure alignment score with hybrid correlation
    padxy = GimageSize - Gimg_shape
    Icorr = hybrid_correlation(img0, img1, padxy=padxy)
    searchScore = Icorr.max()

    return searchScore, a0, a1

def _parallel_search(linearSearch, flagReportProgress, tasks):
    """parallelise the search of linear drfit by creating workers
    """

    linearSearchScore = np.zeros((linearSearch.size, linearSearch.size))

    # create progress bar
    pbar = tqdm(total=linearSearch.size**2, desc='Linear Drift Search',
            leave=False, disable=not flagReportProgress)

    with Pool(cpu_count()) as pool:
        for ret in pool.imap_unordered(_makeimage, tasks, chunksize=1):
            searchScore, a0, a1 = ret
            linearSearchScore[a0, a1] = searchScore

            # update progress
            pbar.update(1)

    # close progress bar
    pbar.close()

    return linearSearchScore

def _serial_search(sm, linearSearch, flagReportProgress, inds, xDrift, yDrift):
    """serial search of linear drfit
    """

    linearSearchScore = np.zeros((linearSearch.size, linearSearch.size))

    # create progress bar for nested for loop
    pbar = tqdm(total=linearSearch.size**2, desc='Linear Drift Search',
                leave=False, disable=not flagReportProgress)

    for a0 in range(linearSearch.size):
        for a1 in range(linearSearch.size):
            # calculate time dependent linear drift
            xyShift = np.hstack([inds*xDrift[a0,a1], inds*yDrift[a0,a1]])

            # apply linear drift to first two images
            # linear drift should be same for all images
            sm.scanOr[:2, ...] += xyShift.T

            # generate trial images after applying specific drifts
            sm = SPmakeImage(sm, 0)
            sm = SPmakeImage(sm, 1)

            # measure alignment score with hybrid correlation
            img0 = sm.imageTransform[0, ...]
            img1 = sm.imageTransform[1, ...]
            padxy = sm.imageSize - sm.img_shape
            Icorr = hybrid_correlation(img0, img1, padxy=padxy)
            linearSearchScore[a0, a1] = Icorr.max()

            # restore the first two image by removing applied linear drift
            sm.scanOr[:2, ...] -= xyShift.T

            # update progress
            pbar.update(1)

    # close the progress bar of the nested for loop
    pbar.close()

    return linearSearchScore

def _shift_origins(sm, dxy):
    """shift origins of the rows by dxy and remake images, where dxy is an
    array contains x and y shifts for every image in sMerge, could be the same
    for every origin or different

    dxy must be broadcastable to scanOr
    """

    # shift the origins
    dxy = np.asarray(dxy)
    sm.scanOr += dxy

    # regenerate every image
    for k in range(sm.numImages):
        sm = SPmakeImage(sm, k)

    return

def _get_initial_alignment(sm):
    """Estimate initial alignment dxy
    """

    # first image set at (0, 0)
    dxy = np.zeros((sm.numImages, 2))

    for k in range(1, sm.numImages):
        # measure alignment score with hybrid correlation
        imgA = sm.imageTransform[k-1, ...]
        imgB = sm.imageTransform[k, ...]
        padxy = sm.imageSize - sm.img_shape
        Icorr = hybrid_correlation(imgA, imgB, padxy=padxy)
        dx, dy = np.unravel_index(Icorr.argmax(), Icorr.shape)

        # check if it wraps over
        nr, nc = Icorr.shape
        dx = (dx + nr/2) % nr - nr/2
        dy = (dy + nc/2) % nc - nc/2

        # alignment relative the one the was correlated
        dxy[k, :] = dxy[k-1, :] + np.array([dx, dy])

    # normalise??
    dxy[:, 0] -= dxy[:, 0].mean()
    dxy[:, 1] -= dxy[:, 1].mean()

    return dxy

def _plot(sm):
    """Plot images and their density after linear drift correction, also
    showing the reference point for later alignment step
    """
    # Plot results, image with scanline origins overlaid
    imagePlot = sm.imageTransform.mean(axis=0)
    dens = sm.imageDensity.prod(axis=0)

    # Scale intensity of image
    mask = dens > 0.5
    imagePlot -= imagePlot[mask].mean()
    imagePlot /= np.sqrt(np.mean(imagePlot[mask]**2))

    fig, ax = plt.subplots()
    ax.matshow(imagePlot, cmap='gray')

    # RGB colours
    cvals = np.array([[1, 0, 0],
                      [0, 0.7, 0],
                      [0, 0.6, 1],
                      [1, 0.7, 0],
                      [1, 0, 1],
                      [0, 0, 1]])

    # put origins on plot
    for k in range(sm.numImages):
        x = sm.scanOr[k, 1, :]
        y = sm.scanOr[k, 0, :]
        c = cvals[k % cvals.shape[0], :]

        ax.plot(x, y, marker='.', markersize=12, linestyle='None', color=c)

    # put a nice cross at the reference point
    ax.plot(sm.ref[1], sm.ref[0], marker='+',
            markersize=20, color=(1,1,0), markeredgewidth=5)
    ax.plot(sm.ref[1], sm.ref[0], marker='+',
            markersize=20, color='k', markeredgewidth=3)

    return
